drivers/openstack/provider.go

```diff
@@ -0,0 +1,63 @@
+// Copyright 2018 Drone.IO Inc
+// Use of this software is governed by the Business Source License
+// that can be found in the LICENSE file.
+
+package openstack
+
+import (
+	"sync"
+	"text/template"
+
+	"github.com/drone/autoscaler"
+	"github.com/drone/autoscaler/drivers/internal/userdata"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+)
+
+// provider implements an OpenStack provider
+type provider struct {
+	init sync.Once
+
+	key      string
+	region   string
+	image    string
+	flavor   string
+	subnet   string
+	pool     string
+	userdata *template.Template
+	groups   []string
+	metadata map[string]string
+
+	computeClient *gophercloud.ServiceClient
+}
+
+// New returns a new OpenStack provider.
+func New(opts ...Option) autoscaler.Provider {
+	p := new(provider)
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	if p.userdata == nil {
+		p.userdata = userdata.T
+	}
+
+	if p.computeClient == nil {
+		opts, err := openstack.AuthOptionsFromEnv()
+
+		if err != nil {
+			panic(err)
```

Comment:
Please don't panic, look at the Google provider, you can gracefully fail if you return an error

Reply:
Thanks. I was thinking about that, but while setupProvider() returns an error,  provider.New() does not and the calling function always returns nil.  Is it acceptable to change the signature of openstack.New() to return an error in this case?

Reply:
Also. I notice that WithUserDataFile() option panics in all providers.